### ISSUE TITLE
An IterableDataset implementation for chunked data

### ISSUE THREAD

**Issue by hhdri** [pytorch/pytorch#60613]
## 🚀 Feature

An IterableDataset implementation for reading chunked data (data with N points in M chunks where N >> M).

## Motivation

Suppose we have data with size N separated into M chunks (N >> M). The data is too big to fit into RAM entirely. As we don’t have random access to data, I was looking to implement a chunk Dataset that inherits IterableDataset (no random access), which supports multiple workers and partial shuffling.

## Pitch

I want a Dataset that can handle chunks with the support of multiple workers. It should distribute the chunks across its workers. Each worker then should iterate over each of its own chunks until all the chunks are done. Ideally, it should support shuffling the chunks at each run.

## Alternatives

I didn’t find anything, so I tried to implement it myself.

## Additional context

Here's my implementation:

```
import numpy as np
import torch
import pandas as pd


class ChunkDatasetIterator:
    def __init__(self, dataset, file_paths):
        self.path_permutation = np.random.permutation(file_paths)
        self.current_df_index = -1
        self.dataset = dataset

    def __iter__(self):
        return self

    def __next__(self):
        if self.current_df_index == -1:
            if self.current_df_index == len(self.path_permutation) - 1:
                raise StopIteration
            self.current_df_index += 1
            self.current_iterator = self.dataset.make_iterator_from_chunk(self.path_permutation[self.current_df_index])

        try:
            result = next(self.current_iterator)
        except StopIteration:
            if self.current_df_index == len(self.path_permutation) - 1:
                raise StopIteration
            else:
                self.current_df_index += 1
                self.current_iterator = \
                    self.dataset.make_iterator_from_chunk(self.path_permutation[self.current_df_index])
                result = next(self.current_iterator)

        return result


class ChunkDataset(torch.utils.data.IterableDataset):
    def __init__(self, file_paths):
        super(ChunkDataset).__init__()
        self.file_paths = file_paths
        self.dataset_size = self.get_overall_length(file_paths)

    def __len__(self):
        return self.dataset_size

    def __iter__(self):
        worker_info = torch.utils.data.get_worker_info()
        if worker_info is None:
            return ChunkDatasetIterator(self, self.file_paths)
        else:
            return ChunkDatasetIterator(
                self,
                [elem for ind, elem in enumerate(self.file_paths) if (ind % worker_info.num_workers) == worker_info.id]
            )

    def make_iterator_from_chunk(self, file_path):
        raise NotImplementedError

    def get_overall_length(self, file_paths):
        raise NotImplementedError


class ChunkDatasetPandasParquet(ChunkDataset):
    def make_iterator_from_chunk(self, file_path):
        return pd.read_parquet(file_path).sample(frac=1).iterrows()

    def get_overall_length(self, file_paths):
        return sum(len(pd.read_parquet(file_path)) for file_path in file_paths)
```

This ChunkDataset creates a ChunkDatasetIterator for each worker and splits chunks of data across workers. Then each worker shuffles the order of its chunks. For getting an iterator from each chunk, `ChunkDatasetIterator` calls `make_iterator_from_chunk` from its Dataset object. I do it this way so I can customize it just by making a class inherited from ChunkDataset.

Then I created `ChunkDatasetPandasParquet`, which treats the chunks as parquet files, reads them with pandas, and also shuffles inside of each chunk.

This code works very well for my use case. I build my model data with apache spark as parquet files and train my model with `ChunkDatasetPandasParquet` in PyTorch.

Please let me know your thoughts on this feature request and/or implementation. Thank you!

cc @SsnL @VitalyFedyunin @ejguan

**Comment by hhdri** [comment]
I found [this implementation](https://discuss.pytorch.org/t/read-dataset-from-tfrecord-format/16409/16) a shorter way to do the same thing. Instead of making an iterator class and instantiating it for each worker, it uses `yield` for making generators.
I think we should close this issue (because `yield`style is very easy to implement), but I wish I could find [this example](https://discuss.pytorch.org/t/read-dataset-from-tfrecord-format/16409/16) in official documentation because I think it is very useful.
